prow/cmd/deck/trace.go

```diff
@@ -0,0 +1,199 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/kube"
+)
+
+type lineData struct {
+	TimeStr   string `json:"time"`
+	time      time.Time
+	PrNum     int    `json:"pr"`
+	Repo      string `json:"repo"`
+	Org       string `json:"org"`
+	EventGUID string `json:"event-GUID"`
+}
+
+type podLine struct {
+	actual      []byte
+	unmarshaled lineData
+}
+
+// linesByTimestamp sorts pod lines by timestamp.
+// Useful when collecting logs across multiple pods.
+type linesByTimestamp []podLine
+
+// linesByTimestamp implements the sort.Interface interface.
+var _ sort.Interface = linesByTimestamp{}
+
+func (l linesByTimestamp) Len() int      { return len(l) }
+func (l linesByTimestamp) Swap(i, j int) { l[i], l[j] = l[j], l[i] }
+func (l linesByTimestamp) Less(i, j int) bool {
+	return l[i].unmarshaled.time.Before(l[j].unmarshaled.time)
+}
+
+// linesByTimestamp implements the fmt.Stringer interface.
+var _ fmt.Stringer = linesByTimestamp{}
+
+// Return valid json.
```

Comment:
```
// String returns the line data as valid JSON
```

Reply:
This method seems weird -- why are we building JSON with string methods? Maybe a custom seralization would be more appropriate for the `podLine` type?

Reply:
The reason I am building it up like this is that we don't know all of the fields contained in `line.actual` at this level. I unmarshal to `podLine` just to figure out whether the line contains any of the fields we care about and also to do sorting based on timestamps. If we were to marshal `podLine` back into JSON we would lose info. I have added a unit test to verify that we return valid json.

Reply:
Couldn't we keep the unmarshalled bits in https://golang.org/pkg/encoding/json/#RawMessage ?

Reply:
`RawMessage` could potentially help in case we don't know the fields. There is also another issue in our case though. Note that the pod logs we scrape are valid JSON on the line level and *something* needs to take all those lines and turn them into a valid JSON list.

Reply:
OK, yeah, this is a weird case for sure. With the test it LGTM